go/marshal/decode.go

```diff
@@ -217,3 +229,75 @@ func nomsValueDecoder(v types.Value, rv reflect.Value) {
 	}
 	rv.Set(reflect.ValueOf(v))
 }
+
+func sliceDecoder(t reflect.Type) decoderFunc {
+	decoderCache.RLock()
+	d := decoderCache.m[t]
+	decoderCache.RUnlock()
+	if d != nil {
+		return d
+	}
+
+	var decoder decoderFunc
+
+	d = func(v types.Value, rv reflect.Value) {
+		list := v.(types.List)
+		if list.Len() == 0 {
+			rv.Set(reflect.MakeSlice(t, 0, 0))
```

Comment:
Doesn't this change the semantics of a an empty list vs a non-empty list?

1. If the list is empty, the struct has a new empty slice
2. If the list isn't empty, the struct reuses the old slice.

I would have thought (1) is better for both (create a new slice each time), then you can use MakeSlice with the exact right size?

Unless I'm misunderstanding.